File: c/LeetCode_141_LinkedListCycle.c
#include <stdio.h>
#include <stdlib.h>

typedef enum {true, false} bool;


struct ListNode {
    int val;
    struct ListNode *next;
};

/*
bool hasCycle(struct ListNode *head) {
    if (head == NULL || head->next == NULL || head->next->next == NULL) {
        return 0;
    }
   
    struct ListNode *p_1 = head;
    struct ListNode *p_2 = head;

    while (p_1 && p_2 && p_2->next != NULL) {
        p_1 = p_1->next;
        p_2 = p_2->next->next;
        
        if (p_1 == p_2) {
            return true;
        }
    }

    return false;
}
*/

bool hasCycle(struct ListNode *head) {
    if (head == NULL || head->next == NULL) {
        return 0;
    }
   
    struct ListNode *p_1 = head;
    struct ListNode *p_2 = head;

    int inx = 1;

    while (p_2->next != NULL) {
        if (inx % 2 == 0) {
            p_1 = p_1->next;
        }
        p_2 = p_2->next;
        
        if (p_1 == p_2) {
            return true;
        }

        ++inx;
    }

    return false;
}


int main() {

    return 0;
}

File: python/LeetCode_141_LinkedListCycle.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def hasCycle(head) -> bool:
    if not head: return False

    slow = fast = head
    while (slow and fast and fast.next):
        slow = slow.next
        fast = fast.next.next
        if (slow == fast): return True

    return False


nodes = ListNode(3)
nodes.next = ListNode(2)
nodes.next.next = ListNode(1)
nodes.next.next.next = ListNode(0)
nodes.next.next.next.next = nodes.next.next

print(hasCycle(nodes))